src/main/java/httpServer/ListenerAndResponder.java

```diff
@@ -40,46 +40,102 @@ public Socket keepListening() throws IOException {
     }
 
     public void readClientInput(Socket clientSocket) throws IOException {
-
+        System.err.println("Starting to read");
         ClientReadable bufferedReader = clientReaderFactory.makeReader(clientSocket);
-
-        String httpRequest;
-        httpRequest = readAllLines(bufferedReader);
+        HTTPRequest httpRequest;
+        httpRequest = readHeaderAndBodyAndReturnHttpRequest(bufferedReader);
         if (httpRequest == null) {
             StatusCodes statusCodes = StatusCodes.BAD_REQUEST;
             String responseText = statusCodes.httpResponse;
             printServerResponse(responseText, clientSocket);
         }
         if (!httpRequest.equals("")) {
-            HTTPRequest request = RequestParser.parse(httpRequest);
-            System.out.println(httpRequest);
-            String httpResponse = router.getResponse(request);
-            printServerResponse(httpResponse, clientSocket);
+            HTTPResponse httpResponse = router.getResponse(httpRequest);
+            String responseString = httpResponse.toString();
+            printServerResponse(responseString, clientSocket);
         }
 
     }
 
-    private void printServerResponse(String message, Socket clientSocket) throws IOException {
-            ClientWriteable printer = clientWriterFactory.makePrinter(clientSocket);
-            System.out.println(message);
-            printer.println(message);
+    private void printServerResponse(String httpResponse, Socket clientSocket) throws IOException {
+        ClientWriteable printer = clientWriterFactory.makePrinter(clientSocket);
+        System.out.println(httpResponse);
+        printer.print(httpResponse);
+        printer.close();
+    }
+
+    private HTTPRequest readHeaderAndBodyAndReturnHttpRequest(ClientReadable bufferedReader)  {
+
+        HTTPRequestBuilder requestBuilder = new HTTPRequestBuilder();
+        requestBuilder = readHeaders(bufferedReader, requestBuilder);
+        int bodyLength = requestBuilder.getContentLengthInt();
+        StringBuilder body = readBody(bufferedReader, bodyLength);
+        return returnHttpRequest(body, requestBuilder);
 
     }
 
-    private String readAllLines(ClientReadable bufferedReader)  {
-        String line;
-        StringBuilder stringBuilder = new StringBuilder();
+    private HTTPRequestBuilder readHeaders(ClientReadable bufferedReader, HTTPRequestBuilder requestBuilder) {
+
+        String character;
+        StringBuilder headers = new StringBuilder();
+
         try {
-            while ((line = bufferedReader.readLine()) != null)
-            {
-                stringBuilder.append(line);
-                stringBuilder.append(CRLF);
-                if (line.equals("")) { break; }
+            while ((character = bufferedReader.read()) != null) {
+
+                headers.append(character);
+                if (checkIfReachedEndOfHeaders(headers)) {
```

Comment:
On the one hand I'm glad you internalized the "more verbose variable names are clearer" but in this case, I don't think it's actually adding to clarity. Maybe we can just call `checkIfReachedEndOfHeaders` to `atEndOfHeaders()`